### ISSUE TITLE
[automower] planner-next-start does not reflect the correct offset to GMT

### ISSUE THREAD

**Issue by aMU5Ed** [openhab/openhab-addons#17147]
## Expected Behavior
The daylightsavings and GMT time offset of thing "planner-next-start" should should work like channel "last-update"

## Current Behavior
Currently the time of the thing "planner-next-start" is 2 hours after the plan in the Husqvarna app.
The Husqvarna app shows the next start at 9:00 (9am) and the openhab things delivers 11:00 (11am).
In central europe we now have a time offset of 2+ to GMT.

More people have this issue:
 - https://community.openhab.org/t/automower-datetime-different-between-next-start-and-local-time/124235
 - https://community.openhab.org/t/how-to-correctly-format-datetimeoffset-in-item/134802/6

## Possible Solution
Handling of the GMT offset and daylight savings to the "planner-next-start" thing.

## Steps to Reproduce (for Bugs)
1. Create a time plan in the Husqvarna App
2. Check the correct timestamp in openhab channel "planner-next-start" thing.

## Context
The time of the thing is not correctly handled

## Your Environment
Openhab 4.2 release - exists since Openhab 3.1 release

**Comment by lsiepel** [comment]
When looking at the code, it gets a timestamp as `long` from the API and it is converted by the method below to a `ZonedDateTime` respecting your openHAB configuration. 

https://github.com/openhab/openhab-addons/blob/4730f64fbbfacf4c72a03526911fa2e20687c59e/bundles/org.openhab.binding.automower/src/main/java/org/openhab/binding/automower/internal/things/AutomowerHandler.java#L352-L355

The only thing i can think of is that the provided `long` is not the number of milliseconds from 1970-01-01T00:00:00Z, but that would be weird.

**Comment by MikeTheTux** [comment]
Automower API pretends that the timestamp that come from the Server are in UTC, while the one that come from the Mower are in Local Mower TimeZone.
The binding treats all time stamps in the same way.
I tried to extend the function above using a ZoneId parameter (UTC or local, e.g. CET). It did not brought the expected result. I think there is something fishy in the way how the timestamps of the API are interpreted.

> Unfortunately the API does not expose the timezone the mower is in.
Can this be looked up from the GPS coordinates the mower is sending? But even in middle Europa, a mower might run with UTC settings ...

**Comment by MikeTheTux** [comment]
Via trial and error (not strictly following the API documentation), I came to following implementation that provides valid results for me:

```
    private ZonedDateTime toZonedDateTime(long timestamp, ZoneId zoneId) {
        logger.warn("ZoneId: {}, timeZoneProvider: {}", zoneId.toString(), timeZoneProvider.getTimeZone());

        Instant timestampInstant = Instant.ofEpochMilli(timestamp);
        ZonedDateTime localTime = ZonedDateTime.ofInstant(timestampInstant, zoneId);
        ZonedDateTime utcTime = ZonedDateTime.ofInstant(timestampInstant, ZoneId.of("UTC"));

        logger.warn("localTime: {}, utcTime: {}", localTime.toString(), utcTime.toString());
        logger.warn("localTime: {}, utcTime: {}", localTime.toEpochSecond(), utcTime.toEpochSecond());

        long timeDiff = localTime.getOffset().getTotalSeconds();
        ZonedDateTime convertedTime = localTime.minusSeconds(timeDiff);
        logger.warn("timeDiff: {}, convertedTime: {}", timeDiff, convertedTime);

        return convertedTime;
    }
```

Log:
```
2024-10-13 11:48:36.649 [WARN ] [wer.internal.things.AutomowerHandler] - CHANNEL_STATUS_LAST_UPDATE: 1728733281198, zoneID(): Europe/Berlin
2024-10-13 11:48:36.650 [WARN ] [wer.internal.things.AutomowerHandler] - ZoneId: Europe/Berlin, timeZoneProvider: Europe/Berlin
2024-10-13 11:48:36.650 [WARN ] [wer.internal.things.AutomowerHandler] - localTime: 2024-10-12T13:41:21.198+02:00[Europe/Berlin], utcTime: 2024-10-12T11:41:21.198Z[UTC]
2024-10-13 11:48:36.651 [WARN ] [wer.internal.things.AutomowerHandler] - localTime: 1728733281, utcTime: 1728733281
2024-10-13 11:48:36.651 [WARN ] [wer.internal.things.AutomowerHandler] - timeDiff: 7200, convertedTime: 2024-10-12T11:41:21.198+02:00[Europe/Berlin]
2024-10-13 11:48:36.658 [WARN ] [wer.internal.things.AutomowerHandler] - CHANNEL_PLANNER_NEXT_START: 1728804600000, zoneID(): Europe/Berlin
2024-10-13 11:48:36.659 [WARN ] [wer.internal.things.AutomowerHandler] - ZoneId: Europe/Berlin, timeZoneProvider: Europe/Berlin
2024-10-13 11:48:36.659 [WARN ] [wer.internal.things.AutomowerHandler] - localTime: 2024-10-13T09:30+02:00[Europe/Berlin], utcTime: 2024-10-13T07:30Z[UTC]
2024-10-13 11:48:36.659 [WARN ] [wer.internal.things.AutomowerHandler] - localTime: 1728804600, utcTime: 1728804600
2024-10-13 11:48:36.660 [WARN ] [wer.internal.things.AutomowerHandler] - timeDiff: 7200, convertedTime: 2024-10-13T07:30+02:00[Europe/Berlin]
```

I expect hat mower time zone is the system timezone.
It seems that CHANNEL_STATUS_LAST_UPDATE and CHANNEL_PLANNER_NEXT_START follow the same logic (not according to docu).

We could add an optional config parameter: time zone of automower. if not set, system time zone is used.

**Comment by lsiepel** [comment]
> We could add an optional config parameter: time zone of automower. if not set, system time zone is used.

yes, that is what I suggested. Not perfect but the best we can do from our position. IMO the api should be changed to support timestamps with unambiguous point in time (always utc or iso-8601 string).

**Comment by MikeTheTux** [comment]
This is the final tested implementation of `toZonedDateTime` according to the automower API: 
```
    private ZonedDateTime toZonedDateTime(long timestamp, ZoneId zoneId) {
        return ZonedDateTime.ofInstant(Instant.ofEpochMilli(timestamp), ZoneId.of("UTC")).withZoneSameLocal(zoneId);
    }
```

In addition an optional `mowerZoneId` config parameter was implemented. I guess we can close this issue thereby.

**Comment by aMU5Ed** [comment]
Will test it when it is part of a release and then close it.
